tokio-trace/src/span.rs

```diff
@@ -0,0 +1,540 @@
+    type Item = (&'a str, &'a dyn Value);
+    type IntoIter = Box<Iterator<Item = (&'a str, &'a dyn Value)> + 'a>; // TODO: unbox
+    fn into_iter(self) -> Self::IntoIter {
+        Box::new(self.fields())
+    }
+}
+
+impl fmt::Debug for Data {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Span")
+            .field("name", &self.inner.name)
+            .field("opened_at", &self.inner.opened_at)
+            .field("parent", &self.parent().unwrap_or(self).name())
+            .field("fields", &self.debug_fields())
+            .field("meta", &self.meta())
+            .finish()
+    }
+}
+
+// ===== impl Inner =====
+
+impl Inner {
+    fn new(data: Data, enter_parent: Option<Span>) -> Arc<Self> {
+        Arc::new(Inner {
+            data,
+            enter_parent,
+            currently_entered: AtomicUsize::new(0)
+        })
+    }
+
+    fn enter(&self, from_state: State) {
+        self.currently_entered
+            .fetch_add(1, Ordering::Release);
+        self.data.set_state(from_state, State::Running);
+    }
+}
+
+impl cmp::PartialEq for Inner {
+    fn eq(&self, other: &Inner) -> bool {
+        self.data == other.data
+    }
+}
+
+impl Hash for Inner {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.data.hash(state);
+    }
+}
+
+// ===== impl DataInner =====
+
+impl Hash for DataInner {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        self.opened_at.hash(state);
+        self.static_meta.hash(state);
+    }
+}
+
+impl cmp::PartialEq for DataInner {
+    fn eq(&self, other: &DataInner) -> bool {
+        self.name == other.name &&
+        self.opened_at == other.opened_at &&
+        self.static_meta == other.static_meta
+    }
+}
+
+#[cfg(test)]
+pub use self::test_support::*;
+#[cfg(test)]
+mod test_support {
+    use ::{ Value, span::State };
+    use std::collections::HashMap;
+
+    pub struct MockSpan {
+        pub name: Option<Option<&'static str>>,
+        pub state: Option<State>,
+        pub fields: HashMap<String, Box<dyn Value>>,
+        // TODO: more
+    }
+
+    pub fn mock() -> MockSpan {
+        MockSpan {
+            name: None,
+            state: None,
+            fields: HashMap::new(),
+        }
+    }
+
+    impl MockSpan {
+        pub fn named(mut self, name: Option<&'static str>) -> Self {
+            self.name = Some(name);
+            self
+        }
+
+        pub fn with_state(mut self, state: State) -> Self {
+            self.state = Some(state);
+            self
+        }
+
+        // TODO: fields, etc
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use ::{subscriber, span};
+    use super::*;
+
+    #[test]
+    fn exit_doesnt_finish_while_handles_still_exist() {
+        // Test that exiting a span only marks it as "done" when no handles
+        // that can re-enter the span exist.
+        let _running = subscriber::mock()
+            .enter(span::mock().named(Some("foo")))
```

Comment:
The more I see these builders, the more I think something like syn's trace proc macro would be helpful.

(I believe we disucssed that before.)

Reply:
Opened #9 to track adding something like that.